### ISSUE TITLE
Traffic stealing takes a very long time to start working

### ISSUE THREAD

**Issue by brookelynne** [metalbear-co/mirrord#2070]
### Bug Description

When I start up my service using the mirrord plugin in GoLand, I immediately see kubernetes health check traffic coming through to my local process, but my manual requests to the pod are served by the pod (not my local process) for the first 5.5 to 6.5 minutes, after which point traffic stealing behaves as expected and I can hit my local breakpoints.

### Steps to Reproduce

1. Enable mirrord plugin.
2. Start debugging and watch local service start up.
3. Send a request to the target deployment and see that it doesn't hit the local breakpoint.
4. Wait 6 minutes (you may continue to send requests every ten or 20 seconds to see when it starts working).
5. Send more requests to the pod after enough time has passed and see that you can now steal traffic.

### Relevant Logs
I don't think these logs are helpful, I just figured I'd put this here because I do see these logs when I make my request before the 5 minute mark, even though my breakpoints don't get hit and the cluster pod responds to the traffic instead.
```shell
2023-11-11T00:38:40.971600Z TRACE ThreadId(01) mirrord_layer::go: c_abi_syscall6_handler: syscall=0 param1=16 param2=824636526592 param3=4096 param4=0 param5=0 param6=0
2023-11-11T00:38:40.971957Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=281 param1=6 param2=140348189827904 param3=128 param4=0 param5=0 param6=0
2023-11-11T00:38:40.972071Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=0 param1=16 param2=824640162689 param3=1 param4=0 param5=0 param6=0
2023-11-11T00:38:40.972089Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=281 param1=6 param2=140348189827904 param3=128 param4=0 param5=0 param6=0
2023-11-11T00:38:40.972086Z TRACE ThreadId(07) mirrord_layer::go: c_abi_syscall6_handler: syscall=281 param1=6 param2=140348181435200 param3=128 param4=0 param5=0 param6=0
2023-11-11T00:38:40.972103Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=281 param1=6 param2=140348189827904 param3=128 param4=220473 param5=0 param6=0
2023-11-11T00:38:40.972282Z TRACE ThreadId(01) mirrord_layer::go: c_abi_syscall6_handler: syscall=1 param1=10 param2=824637767680 param3=70 param4=0 param5=0 param6=0
2023-11-11T00:38:40.982939Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=0 param1=10 param2=824636342272 param3=32768 param4=0 param5=0 param6=0
2023-11-11T00:38:40.983050Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=0 param1=10 param2=824636342272 param3=32768 param4=0 param5=0 param6=0
2023-11-11T00:38:40.983032Z TRACE ThreadId(01) mirrord_layer::go: c_abi_syscall6_handler: syscall=281 param1=6 param2=140723651052648 param3=128 param4=0 param5=0 param6=0
2023-11-11T00:38:40.983130Z TRACE ThreadId(01) mirrord_layer::go: c_abi_syscall6_handler: syscall=1 param1=10 param2=824637767680 param3=17 param4=0 param5=0 param6=0
2023-11-11T00:38:40.983139Z TRACE ThreadId(07) mirrord_layer::go: c_abi_syscall6_handler: syscall=281 param1=6 param2=140348181435200 param3=128 param4=0 param5=0 param6=0
2023-11-11T00:38:40.983182Z TRACE ThreadId(07) mirrord_layer::go: c_abi_syscall6_handler: syscall=281 param1=6 param2=140348181435200 param3=128 param4=220462 param5=0 param6=0
2023-11-11T00:38:40.983213Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=1 param1=16 param2=824638046208 param3=187 param4=0 param5=0 param6=0
2023-11-11T00:38:40.983270Z TRACE ThreadId(08) mirrord_layer::go: c_abi_syscall6_handler: syscall=0 param1=16 param2=824636526592 param3=4096 param4=0 param5=0 param6=0
```

### Your operating system and version

Ubuntu 22.04.3 LTS

### Local process

Golang 1.21

### Additional Info

I'm using the GoLand mirrord plugin, debugging a Go service against a deployment with one pod with one container, using this config:
```json
{
    "target": "deployment/name-of-my-service",
    "feature": {
        "network": {
            "incoming": {
                "mode": "steal"
            },
            "outgoing": true
        },
        "fs": "read",
        "env": false
    },
    "agent": {
        "ephemeral": true
    }
}
```
I'm using ephemeral containers because our cluster is always at capacity and there apparently just isn't room to run an additional pod. (I get `Create agent failed. NodePodLimitExceeded( "mynode01.my.domain.com", 230, )` when I don't use ephemeral.) I'm stealing instead of mirroring because we're running Istio.

I reached out to a coworker to see if it's a cluster problem or a me problem, and I'm going to verify whether this happens with other deployments or just this one. Will follow up if I hear anything interesting.

**Comment by aviramha** [comment]
Thanks for reporting this.
I'd like you to use kubectl debug command to use it on the same pod as your target using mirrord with mirrord-agent image then provide output of 
iptables -nvL -t NAT
and iptables -nvL
Both commands before mirrord, after mirrord and when requests are stolen and before they are as the problem probably lies there.

**Comment by brookelynne** [comment]
Thank you for responding so fast!
I'm trying `kubectl debug -it my-service --image ghcr.io/metalbear-co/mirrord:3.75.1 --target=my-service` and the debug container immediately terminates after outputting the mirrord-agent help. Am I using `kubectl debug` incorrectly?

**Comment by aviramha** [comment]
> Thank you for responding so fast!
> 
> I'm trying `kubectl debug -it my-service --image ghcr.io/metalbear-co/mirrord:3.75.1 --target=my-service` and the debug container immediately terminates after outputting the mirrord-agent help. Am I using `kubectl debug` incorrectly?

Try
`kubectl debug -it my-service --image ghcr.io/metalbear-co/mirrord:3.75.1 --target=my-service sh`

**Comment by brookelynne** [comment]
Adding `sh` gives
```
Targeting container "my-service". If you don't see processes from this container it may be because the container runtime doesn't support this feature.
Defaulting debug container name to debugger-gnn6r.
Run a local process in the context of a cloud environment

Usage: mirrord-agent [OPTIONS] <COMMAND>

Commands:
  targeted    
  ephemeral   Inform the agent to use `proc/1/root` as the root directory
  targetless  
  help        Print this message or the help of the given subcommand(s)

Options:
  -l, --communicate-port <COMMUNICATE_PORT>
          Port to use for communication [default: 61337]
  -t, --communication-timeout <COMMUNICATION_TIMEOUT>
          Communication timeout in seconds [default: 30]
  -i, --network-interface <NETWORK_INTERFACE>
          Interface to use
  -p, --pause
          Pause the target container while clients are connected
      --base-protocol-version <BASE_PROTOCOL_VERSION>
          [default: 1.2.1]
  -h, --help
          Print help
  -V, --version
          Print version
Error from server (NotFound): pods "sh" not found
```

**Comment by infiniteregrets** [comment]
> Adding `sh` gives
> 
> ```
> Targeting container "my-service". If you don't see processes from this container it may be because the container runtime doesn't support this feature.
> Defaulting debug container name to debugger-gnn6r.
> Run a local process in the context of a cloud environment
> 
> Usage: mirrord-agent [OPTIONS] <COMMAND>
> 
> Commands:
>   targeted    
>   ephemeral   Inform the agent to use `proc/1/root` as the root directory
>   targetless  
>   help        Print this message or the help of the given subcommand(s)
> 
> Options:
>   -l, --communicate-port <COMMUNICATE_PORT>
>           Port to use for communication [default: 61337]
>   -t, --communication-timeout <COMMUNICATION_TIMEOUT>
>           Communication timeout in seconds [default: 30]
>   -i, --network-interface <NETWORK_INTERFACE>
>           Interface to use
>   -p, --pause
>           Pause the target container while clients are connected
>       --base-protocol-version <BASE_PROTOCOL_VERSION>
>           [default: 1.2.1]
>   -h, --help
>           Print help
>   -V, --version
>           Print version
> Error from server (NotFound): pods "sh" not found
> ```

Hi @brookelynne, pitching in on Aviram's behalf, please add an argument separator and it should work fine:
```
kubectl debug -it my-service --image ghcr.io/metalbear-co/mirrord:3.75.1 --target=my-service -- sh
```

**Comment by brookelynne** [comment]
Success! Thank you, @infiniteregrets !
```
# iptables -nvL -t NAT
iptables v1.8.9 (legacy): can't initialize iptables table `NAT': Permission denied (you must be root)
Perhaps iptables or your kernel needs to be upgraded.
```
When I watch the pod logs as I start up my local process with mirrord, I get this:
```
+ my-service-65d858d9db-kktm6 › mirrord-agent-gnlzeaktom
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom agent ready - version 3.75.1
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom 2023-11-15T23:35:51.328135Z  WARN ThreadId(05) mirrord_agent::steal::ip_tables::flush_connections: `conntrack` output is Output {
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom     status: ExitStatus(
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom         unix_wait_status(
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom             256,
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom         ),
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom     ),
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom     stdout: "",
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom     stderr: "conntrack v1.4.7 (conntrack-tools): 0 flow entries have been updated.\n",
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom }
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom 2023-11-15T23:35:51.790489Z  WARN ThreadId(07) mirrord_agent::outgoing::udp: interceptor_task -> no messages left
my-service-65d858d9db-kktm6 mirrord-agent-gnlzeaktom 2023-11-15T23:35:51.790469Z  WARN ThreadId(06) mirrord_agent::outgoing: interceptor_task -> no messages left
```
I waited long enough for the stealing to start working and didn't see any additional logs between startup and then.

I asked a couple of my coworkers if they were experiencing this same latency when running mirrord and they both said no, which leads me to believe it is not a cluster issue. They're both on Mac, for what it's worth.

**Comment by infiniteregrets** [comment]
Thanks for the info! @brookelynne do you mind running kubectl debug with the flag `--profile=netadmin`?

A small mistake though, chains are case sensitive, so you'd have to run the following `iptables -nvL -t nat`

Thanks again and sorry I forgot to give you the correct flag (:

**Comment by aviramha** [comment]
In addition to what @infiniteregrets suggested, I'm assuming your coworkers are targeting the same deployment?

**Comment by brookelynne** [comment]
@aviramha Yes, I just confirmed today. I asked one of my coworkers to target the exact same pod I've tried to target, using the exact same mirrord config and debugger config, and it began stealing traffic immediately for him but not for me.

@infiniteregrets Thank you for the remote troubleshooting!
For context, port 2112 is our health check, which forwards just fine right from the beginning. Port 50052 is the one I'm interacting with where the traffic steal takes a long time to start working. (The traffic comes in on 50052, then gets forwarded to 50051 through some middleware stack I'm not very familiar with. My local process is listening on all of these.)

Before mirrord:
```
# iptables -nvL -t nat
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain INPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination     
#
#
# iptables -nvL
Chain INPUT (policy ACCEPT 1300 packets, 103K bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 1263 packets, 213K bytes)
 pkts bytes target     prot opt in     out     source               destination        
```
After mirrord, before interacting with my service:
```
# iptables -nvL -t nat
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
    2   120 MIRRORD_INPUT_fuRJd  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain INPUT (policy ACCEPT 2 packets, 120 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 12 packets, 1087 bytes)
 pkts bytes target     prot opt in     out     source               destination         
   14  1245 MIRRORD_STANDARD_aL1bX  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain POSTROUTING (policy ACCEPT 12 packets, 1087 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain MIRRORD_INPUT_fuRJd (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:50051 redir ports 43569
    0     0 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:6060 redir ports 43569
    0     0 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:50052 redir ports 43569
    2   120 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:2112 redir ports 43569
    0     0 RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain MIRRORD_STANDARD_aL1bX (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    3   221 RETURN     6    --  *      *       0.0.0.0/0            0.0.0.0/0            owner GID match 4395
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:50051 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:6060 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:50052 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:2112 redir ports 43569
   11  1024 RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           
#
#
# iptables -nvL
Chain INPUT (policy ACCEPT 338 packets, 33192 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  359 34404 MIRRORD_INPUT_SGDgR  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 352 packets, 40919 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain MIRRORD_INPUT_SGDgR (1 references)
 pkts bytes target     prot opt in     out     source               destination         
   21  1212 REJECT     6    --  *      *       0.0.0.0/0            0.0.0.0/0            connmark match  0x1 reject-with tcp-reset
  338 33192 RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           
```

After interacting with my service (traffic steal failed):
```
# iptables -nvL -t nat
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
   32  1912 MIRRORD_INPUT_fuRJd  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain INPUT (policy ACCEPT 32 packets, 1912 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 14 packets, 1191 bytes)
 pkts bytes target     prot opt in     out     source               destination         
   17  1431 MIRRORD_STANDARD_aL1bX  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain POSTROUTING (policy ACCEPT 15 packets, 1273 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain MIRRORD_INPUT_fuRJd (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:50051 redir ports 43569
    0     0 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:6060 redir ports 43569
    1    60 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:50052 redir ports 43569
   31  1852 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:2112 redir ports 43569
    0     0 RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain MIRRORD_STANDARD_aL1bX (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    3   221 RETURN     6    --  *      *       0.0.0.0/0            0.0.0.0/0            owner GID match 4395
    1    82 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:50051 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:6060 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:50052 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:2112 redir ports 43569
   13  1128 RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           
#
#
# iptables -nvL
Chain INPUT (policy ACCEPT 744 packets, 117K bytes)
 pkts bytes target     prot opt in     out     source               destination         
  765  118K MIRRORD_INPUT_SGDgR  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 749 packets, 135K bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain MIRRORD_INPUT_SGDgR (1 references)
 pkts bytes target     prot opt in     out     source               destination         
   21  1212 REJECT     6    --  *      *       0.0.0.0/0            0.0.0.0/0            connmark match  0x1 reject-with tcp-reset
  744  117K RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           
```

After waiting 6 minutes and interacting with my service again (steal succeeded):
```
# iptables -nvL -t nat
Chain PREROUTING (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         
  120  7192 MIRRORD_INPUT_fuRJd  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain INPUT (policy ACCEPT 120 packets, 7192 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 34 packets, 3091 bytes)
 pkts bytes target     prot opt in     out     source               destination         
   37  3331 MIRRORD_STANDARD_aL1bX  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain POSTROUTING (policy ACCEPT 35 packets, 3173 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain MIRRORD_INPUT_fuRJd (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    0     0 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:50051 redir ports 43569
    0     0 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:6060 redir ports 43569
    3   180 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:50052 redir ports 43569
  117  7012 REDIRECT   6    --  *      *       0.0.0.0/0            0.0.0.0/0            tcp dpt:2112 redir ports 43569
    0     0 RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain MIRRORD_STANDARD_aL1bX (1 references)
 pkts bytes target     prot opt in     out     source               destination         
    4   281 RETURN     6    --  *      *       0.0.0.0/0            0.0.0.0/0            owner GID match 4395
    1    82 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:50051 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:6060 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:50052 redir ports 43569
    0     0 REDIRECT   6    --  *      lo      0.0.0.0/0            0.0.0.0/0            tcp dpt:2112 redir ports 43569
   32  2968 RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0           
# 
# 
# iptables -nvL
Chain INPUT (policy ACCEPT 3151 packets, 473K bytes)
 pkts bytes target     prot opt in     out     source               destination         
 3172  474K MIRRORD_INPUT_SGDgR  0    --  *      *       0.0.0.0/0            0.0.0.0/0           

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain OUTPUT (policy ACCEPT 3075 packets, 607K bytes)
 pkts bytes target     prot opt in     out     source               destination         

Chain MIRRORD_INPUT_SGDgR (1 references)
 pkts bytes target     prot opt in     out     source               destination         
   21  1212 REJECT     6    --  *      *       0.0.0.0/0            0.0.0.0/0            connmark match  0x1 reject-with tcp-reset
 3151  473K RETURN     0    --  *      *       0.0.0.0/0            0.0.0.0/0    
```
At least within the pod, it looks like the mirrord-agent is doing what it's supposed to...

**Comment by aviramha** [comment]
The iptables seem fine at first glance, can you share the agent logs?
Just to verify, you're not having multiple mirrord sessions against the same target at the same time, right?
Also, we're talking about incoming traffic - not queues/streams right?

**Comment by brookelynne** [comment]
Sorry for the late response; I had two last-minute trips come up last week.
Correct, we're not concurrently using mirrord against the same target. We tested consecutively to verify that it was an issue with my local machine rather than with the pod or cluster.
Yes, this is just incoming HTTP traffic. In this instance we handle REST on 50052 and convert it to a GRPC request against port 50051.

While gathering agent logs today, I noticed that the traffic stealing began working much quicker than before (maybe 2 minutes instead of 5). I restarted the process in order to time it and that second time the stealing worked immediately. I think it's possible my plugin version was out of date, and a fresh boot of my computer with a fresh load of GoLand and update of plugins solved the issue. I apologize profusely for not checking the plugin version before filing a bug report, and for wasting your time. Thank you so much for your unbelievably fast responses. You guys are the best!
